File: Medium/C++/Subsets_II(90).cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>>res;
        sort(nums.begin(), nums.end());
        vector<int>comb;
        backtrack(res, nums, 0, comb);
        return res;
    }
private:
    void backtrack(vector<vector<int>>& res, vector<int>& nums, int pos, vector<int>& comb){
        res.push_back(comb);
        for(int i = pos; i < nums.size(); i++){
            if(i == pos || nums[i] != nums[i - 1]){
                comb.push_back(nums[i]);
                backtrack(res, nums, i + 1, comb);
                comb.pop_back();
            }
        }
    }
};

File: Medium/Java/Subsets_II(90).java
class Solution {
    private List<List<Integer>> ans;
    private int[] nums;
    
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        this.ans = new ArrayList<>();
        this.nums = nums;
        
        Arrays.sort(nums);
        backtrack(0, new ArrayList<>());
        return ans;
    }
    
    // Given subset {... nums[i-1]} as prefix
    // Consider all unique subsets from nums[i ... N-1]
    private void backtrack(int i, List<Integer> prefix) {
        ans.add(new ArrayList<>(prefix));
        
        for (int j = i; j < nums.length; j++) {
            if (j == i || nums[j] != nums[j - 1]) {
                prefix.add(nums[j]);
                backtrack(j + 1, prefix);
                prefix.remove(prefix.size() - 1);
            }
        }
    }
}